design-proposals/klip-10-suppress.md

```diff
@@ -1 +1,123 @@
+    GROUP BY server
+    EMIT CHANGES;
+```
+
+We have a downstream microservice that consumes from the `alerts` topic and sends an email whenever
+the count is less than 3 for a server (i.e. a server missed 2 out of 5 heartbeat healthchecks in a
+window of 5 minutes). 
+
+Today, building such an application would not work the way you want it to! The downstream service
+would get intermediate results (the exact results are slightly unpredictable, as they depend on
+the `commit.interval.ms` and `cache.max.bytes.buffering` configurations), and would spuriously
+alert. For example, if the `healtcheck` stream had the following data:
+```
+ts: min_1 server: abc
+ts: min_2 server: abc
+ts: min_3 server: abc
+```
+the `alerts` topic would get the update three times (each time with incrementing `count` column 
+value), and would send three emails when none should be sent.
+
+## Scope
+
+**In Scope**: 
+- Outline the features of suppress that need to be represented in KSQL
+- Syntax for suppress functionality
+- Outline of potential implementation
+
+**Out of Scope**:
+- Grace periods for windows
+- Custom suppressions
+
+## Value/Return
+
+This feature enables functionality in KSQL that is not possible today, and provides stricter
+semantic guarantees on KSQL aggregation queries.
+
+## Public APIS
+
+The public API will extend the `EMIT` clause to support refinement operators other than `CHANGES`
+(which is the only currently supported refinement). The proposed syntax is:
+
+```sql
+SELECT select_expression
+    FROM source
+    WINDOW window_expression
+    GROUP BY grouping_expression
+    EMIT refinement_expression
+```
+
+A `refinement_expression` has the following syntax:
+```sql
+CHANGES | FINAL [EAGERLY] [WHEN suppression]
+```
+
+And behaves in the following ways:
+- If `CHANGES` is specified, then all intermediate changes will be materialized
+- If `FINAL` is specified, then output will be suppressed depending on the `suppression`:
+    - if `EAGERLY` is specified, the suppression is a best effort attempt to reduce duplicate output
+      data, but no guarantees are made if a constraint is met (e.g. buffer fills up)
+    - if `EAGERLY` is omitted, then the query will fail if the buffer fills up before the refinement
+      condition is met
+
+A `suppression` will implement `Suppressed`. KSQL will provide two builtin suppression mechanisms,
+with `WINDOW_CLOSED()` being the default if no suppression is supplied:
+- `WINDOW_CLOSED()` will create a suppression policy that emits events when the window is closed
+- `TIME_LIMIT(time_expression)` will create a suppression policy that emits events when a certain
+  time period has passed since seeing a given key for the first time
+
+We will model the suppression as method calls in order to allow flexibility in the syntax if
+different types of suppressions are added in the future. We could also extend this to allow user
+implemented suppressions in the same way that we support UDFs.
```

Comment:
Interesting idea -- this is not supported in Kafka Streams atm, and it's unclear to me who this would work? It's not part of this proposal, but as you outline the idea I am wondering how this could look like?

Reply:
Ah I assumed that anyone who implemented the `Suppressed` interface could plug into the interface. If that's not the intention then maybe I can simplify the syntax - but I wanted to leave it open in case that was the idea.

Reply:
Makes sense. For clarity, people don't implement `Suppressed`. In fact, this is expressly forbidden. Instead, they use the static methods on `Suppressed` to configure the operation.

Reply:
@vvcephei is there any intent to open this up in the future?